Source/WTF/wtf/CompactRefPtr.h

```diff
@@ -0,0 +1,193 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <utility>
+#include <wtf/RawPtrTraits.h>
+#include <wtf/RefPtr.h>
+#include <wtf/StdLibExtras.h>
+
+namespace WTF {
+
+#if CPU(ADDRESS64)
+#if CPU(ARM64) && OS(DARWIN)
+#if MACH_VM_MAX_ADDRESS_RAW < (1ULL << 36)
+#define HAVE_36BIT_ADDRESS 1
+#endif
+#endif
+#endif // CPU(ADDRESS64)
+
+template <typename T>
+class Compacted {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+#if HAVE(36BIT_ADDRESS)
+    static_assert(alignof(T) >= 16);
+    using StorageSize = uint32_t;
+#else
+    using StorageSize = uintptr_t;
+#endif
+    static constexpr bool isCompactedType = true;
+
+    Compacted()
+    {
+        set(nullptr);
+    }
+
+    Compacted(nullptr_t)
+    {
+        set(nullptr);
+    }
+
+    Compacted(T* ptr)
+    {
+        set(ptr);
+    }
+
+    Compacted<T>& operator=(T* ptr)
+    {
+        set(ptr);
+        return *this;
+    }
+
+    T* operator->() const { return get(); }
+
+    T& operator*() const { return *get(); }
+
+    bool operator!() const { return !get(); }
+
+    explicit operator bool() const { return !!get(); }
+
+    T* get() const
+    {
+        return decode(m_ptr);
+    }
+
+    void set(T* ptr)
+    {
+        m_ptr = encode(ptr);
+    }
+
+    template <class U>
+    T* exchange(U&& newValue)
+    {
+        T* oldValue = get();
+        set(std::forward<U>(newValue));
+        return oldValue;
+    }
+
+    void swap(nullptr_t) { set(nullptr); }
+
+    void swap(Compacted& other)
+    {
+        T* t1 = get();
+        T* t2 = other.get();
+        set(t2);
+        other.set(t1);
+    }
+
+    template <typename Other, typename = std::enable_if_t<Other::isCompactedType>>
+    void swap(Other& other)
+    {
+        T* t1 = get();
+        T* t2 = other.get();
+        set(t2);
+        other.set(t1);
+    }
+
+    void swap(T* t2)
+    {
+        T* t1 = get();
+        std::swap(t1, t2);
+        set(t1);
+    }
+
+    static ALWAYS_INLINE constexpr StorageSize encode(uintptr_t ptr)
+    {
+#if HAVE(36BIT_ADDRESS)
+        ASSERT_UNDER_CONSTEXPR_CONTEXT(!(ptr & 0xF));
+        return static_cast<StorageSize>(ptr >> 4);
```

Comment:
Rather than using literals like 4, and 0xF, can you define some constants so that every place these are used, we can know that they are that value for a common reason?  e.g. define these in the CompactPtr class"
```
static constexpr bitsShift = 4;
static constexpr uintptr_t alignmentMask = (1ull << bitsShift) - 1;
```

Also, if you use a literal and accidentally type 3 or 5 instead of 4 (due to a typo), the compiler will happily compile it, and you will have to debug the fall out.  But if you used a symbolic constant like `CompactPtr::bitsShift` and happen to have a typo, the compiler will likely catch your error for you with a build time error.  So, in general, it's beneficial to use symbolic constants instead of literals.

Reply:
Updated